vioscsi/vioscsi.c

```diff
@@ -438,60 +438,189 @@ VioScsiFindAdapter(
+         * a "break glass" configuration override knob in the Windows registry.
+         *
+         * The maximum transfer length can be adjusted by creating and editing
+         * the following Registry key, which the code below will read and apply.
+         * [HKEY_LOCAL_MACHINE\SYSTEM\CurrentControlSet\Services\vioscsi\Parameters\Device]
+         * "PhysicalBreaks"={dword value here}
+         *
+         * TODO: Update this logic to incorporate indirect descriptor queue
+         * size limits if/when it is fully ratified into the virtio spec.
+         * This will allow the guest and host to negotiate on this specific
+         * limit. Reference:
+         *   https://github.com/oasis-tcs/virtio-spec/issues/122
+         */
+        adaptExt->indirect = CHECKBIT(adaptExt->features, VIRTIO_RING_F_INDIRECT_DESC);
         if (adaptExt->indirect) {
+            /* First take the default MAX_PHYS_SEGMENTS, then inspect the
+             * registry to see if the user has provided an override.
+             * If an override is configured, VioScsiReadRegistry will modify
+             * max_physical_breaks.
+             */
             adaptExt->max_physical_breaks = MAX_PHYS_SEGMENTS;
             VioScsiReadRegistry(DeviceExtension);
         }
+
+        /* Reminder, ConfigInfo is the sharded data structure that Windows is
+         * using, so we need to set physical breaks using the logic we've
+         * cached in our private adapter extention state.
+         *
+         * Note: Having one extra page allows Windows to do some optimizations
+         * to do memory merging, such that the sgList will hopefully contain
+         * less overall buffers, which will be less for the host backend to
+         * process.
+         */
         ConfigInfo->NumberOfPhysicalBreaks = adaptExt->max_physical_breaks + 1;
-        if (adaptExt->scsi_config.max_sectors > 0 && adaptExt->scsi_config.max_sectors != 0xFFFF) {
-            ConfigInfo->NumberOfPhysicalBreaks = min (ConfigInfo->NumberOfPhysicalBreaks, adaptExt->scsi_config.max_sectors);
-            adaptExt->max_physical_breaks = ConfigInfo->NumberOfPhysicalBreaks - 1;
+        ConfigInfo->MaximumTransferLength  = adaptExt->max_physical_breaks * PAGE_SIZE;
+
+        /* When we get to this point in the initialization, we need to make
+         * sure we didn't accidentally set the transfer size higher than the
+         * max_sector restricts, as some platforms may act strange (or even
+         * error/fail) if their max_sectors limits are violated.
+         */
+        if (adaptExt->scsi_config.max_sectors > 0 &&
+            adaptExt->scsi_config.max_sectors != 0xFFFF &&
+            ConfigInfo->MaximumTransferLength > (adaptExt->scsi_config.max_sectors * SECTOR_SIZE)) {
+
+            ConfigInfo->MaximumTransferLength  = adaptExt->scsi_config.max_sectors * SECTOR_SIZE;
+            adaptExt->max_physical_breaks      = ConfigInfo->MaximumTransferLength / PAGE_SIZE;
```

Comment:
I think it would be less clumsy to replace this whole block with:
```
int mbp_ms_clamp = adaptExt->scsi_config.max_sectors * SECTOR_SIZE / PAGE_SIZE;
if (mbp_ms_clamp > 0)
    adaptExt->max_physical_breaks = min(mbp_ms_clamp, adaptExt->max_physical_breaks);
```
and put it after the registry reading but before NOPB and MTL setting.

The `!= 0xffff` check is unnecessary and bad anyway.

Reply:
yea, we need that because vhost-scsi.c in qemu sets max_sectors to 0xFFFF as default, which is 65k, so yes it looks clunky, but it serves a purpose.